src/v/archival/ntp_archiver_service.cc

```diff
@@ -1515,6 +1511,16 @@ ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
         };
     }
 
+    vassert(
```

Comment:
Can we replace this (together with the above if statements) with `ss::visit` to have compile time check that we're not missing any variants?

Reply:
I see there is a ton of code below. It will need to be extracted into a separate function to make it readable. Potentially follow up PR?

Reply:
I changed this to use `ss::visit` but it was a little tricky, as `ss::visit` doesn't seem to work well with rvalue references (`upload_candidate_with_locks&&`) complaining of non-exhaustive visitors, so I had to add helper methods and use `auto&&`. 

I'll hold off on making a change for the adjacent segment merger though, maybe in a follow up. Maybe there is a better way rather than defining overloads for all variant types.

Reply:
Why do you need rvalue references? I'd assume a simple references will work just fine. You have examples in tests, [like this one](https://github.com/redpanda-data/redpanda/pull/15089/files#diff-d61d3bb2f3d53ce97e9082898f81574487d5e8503692942486057cfc6f32609aR139).

You can move out of a reference too. Not sure I understand the need with custom visitors. Much harder to read too.

Reply:
Tried this and don't get any errors:

```
    co_return co_await ss::visit(
      candidate_result,
      [](archival::candidate_creation_error&) -> ss::future<scheduled_upload> {
          co_return scheduled_upload{};
      },
      [](archival::skip_offset_range&) -> ss::future<scheduled_upload> {
          co_return scheduled_upload{};
      },
      [](archival::upload_candidate_with_locks& c)
        -> ss::future<scheduled_upload> { co_return scheduled_upload{}; });
```

Reply:
Yes, moving out of a reference seems better, I wanted to use rvalue references as segment locks are non-copyable but using a reference and moving out of it seems to work around that. Changed to use normal references.